dns/client.go

```diff
@@ -0,0 +1,114 @@
+package dns
+
+import (
+	"context"
+	"fmt"
+	"net"
+
+	"github.com/miekg/dns"
+)
+
+// Resolver is the interface that wraps the Resolve method.
+//
+// Resolve resolves a domain name to an IP address. It returns a slice of IP
+// addresses as strings.
+type Resolver interface {
+	Resolve(ctx context.Context, query, recordType string, nameserver Nameserver) ([]string, error)
+}
+
+// Lookuper is the interface that wraps the Lookup method.
+//
+// As opposed to a Resolver which uses a specific nameserver to resolve the
+// query, a Lookuper uses the system's default resolver.
+//
+// Lookup resolves a domain name to an IP address. It returns a slice of IP
+// addresses as strings.
+type Lookuper interface {
+	Lookup(ctx context.Context, hostname string) ([]string, error)
+}
+
+// Client is a DNS resolver that uses the `miekg/dns` package under the hood.
+//
+// It implements the Resolver interface.
+type Client struct {
+	// client is the DNS client used to resolve queries.
+	client dns.Client
+}
+
+// Ensure our Client implements the Resolver interface
+var _ Resolver = &Client{}
+
+// Ensure our Client implements the Lookuper interface
+var _ Lookuper = &Client{}
```

Comment:
They sound to be unused. Why do we need them?

Reply:
The intention was to make the Client somewhat generic enough to be usable in the context described in the new HTTP module design doc: having the ability to pass in a DNS module object/function, or a (future) TCP module object/function to the HTTP client's constructor.

We encountered a similar case recently in the context of streams/fs, and having the ability to reuse underlying module's behavior as interfaces proved useful.

But I'm happy to drop those if you prefer as I acknowledge those are more "future guarantees", rather than needed immediately 👍🏻